app/src/contexts/surveyContext.tsx

```diff
@@ -0,0 +1,62 @@
+import { useBiohubApi } from 'hooks/useBioHubApi';
+import useDataLoader, { DataLoader } from 'hooks/useDataLoader';
+import { IGetSurveyForViewResponse } from 'interfaces/useSurveyApi.interface';
+import React, { createContext, PropsWithChildren, useContext, useEffect } from 'react';
+import { useParams } from 'react-router';
+
+export interface ISurveyContext {
+  surveyDataLoader: DataLoader<[project_id: number, survey_id: number], IGetSurveyForViewResponse, unknown>;
+  projectId: number | null;
+  surveyId: number | null;
+  _setSurveyId(surveyId: number): void;
+  _setProjectId(projectId: number): void;
+}
+
+const SurveyContext = createContext<ISurveyContext>({
+  surveyDataLoader: {} as DataLoader<[project_id: number, survey_id: number], IGetSurveyForViewResponse, unknown>,
+  projectId: null,
+  surveyId: null,
+  _setSurveyId: () => null,
+  _setProjectId: () => null
+});
+
+export const SurveyContextProvider = (props: PropsWithChildren<Record<never, any>>) => {
+  const [surveyId, setSurveyId] = React.useState<number | null>(null);
+  const [projectId, setProjectId] = React.useState<number | null>(null);
+
+  const biohubApi = useBiohubApi();
+  const surveyDataLoader = useDataLoader(biohubApi.survey.getSurveyForView);
+
+  useEffect(() => {
+    if (projectId && surveyId) {
+      surveyDataLoader.refresh(projectId, surveyId);
+    }
+  }, [projectId, surveyDataLoader, surveyId]);
+
+  const surveyContext: ISurveyContext = {
+    surveyDataLoader,
+    projectId,
+    surveyId,
+    _setSurveyId: setSurveyId,
+    _setProjectId: setProjectId
+  };
+
+  return <SurveyContext.Provider value={surveyContext} {...props} />;
+};
+
+/**
+ *
+ * @returns
+ */
+export const useSurveyContext = (): ISurveyContext => {
+  const urlParams = useParams();
+  const surveyContext = useContext(SurveyContext);
+
+  // Update survey each time the survey ID or project ID changes
+  useEffect(() => {
+    surveyContext._setSurveyId(urlParams['survey_id']);
+    surveyContext._setProjectId(urlParams['id']);
+  }, [surveyContext, urlParams]);
```

Comment:
Could/should this be `survey_id` and `id` directly, rather than urlParms?

Reply:
My thinking was that rather than having Survey Context consumers poke at the URL params, just let the Survey Context fetch them and update the project and survey ID, so subscribers can just reliably get these two IDs from the survey context (assuming that the survey context is only used when these URL params are made available by a router higher up in the state tree)

Reply:
I just mean as far as the useEffect dependencies are concerned. If `urlParams` is included, then other url params outside of project/survey id would cause the useEffect to re-run.

That being said, its quite possible that when `useParams` changes, the whole context is going to update regardless, so maybe narrowing the useEffect dependencies wouldn't help anything. Not sure though without testing it.

Reply:
Ohh I totally misunderstood. Now I'm thinking maybe it should just be `[urlParams['survey_id'], urlParams['id']]`. If we include `surveyContext`, then it will run twice - once when `urlParams` updates, and again when `surveyContext.surveyId` changes in the parent...